### ISSUE TITLE
Change default cfg file name to be more generic?

### ISSUE THREAD

**Issue by billsacks** [ESMCI/manage_externals#52]
Even if manage_externals is used exclusively in CESM and its components, having a default name of CESM.cfg doesn't always make sense. For example, @mnlevy1981 is planning to use this in place of svn externals in POP to pull in POP's dependencies (cvmix, marbl). I assume he's going to want to be able to run checkout_externals from a POP checkout to get these. Currently this would require a top-level CESM.cfg file, which doesn't really make sense because he wouldn't be pulling in other CESM components.

What about something generic like EXTERNALS.cfg?

**Comment by billsacks** [comment]
@bandre-ucar what do you think about this?

**Comment by bandre-ucar** [comment]
I don't have a preference, beyond if we make this change, it should happen before v1.0, and that if we are going to move away from using the acronyms we should drop the all caps and just use Externals.cfg

Some things to consider and then let me know how you want to proceed:

We have a 'default' for checkout_externals, but that is only a convenience. The user can override it by using the --externals option.

Things start to get weird when you look at this from the context of the various nested repositories.

There are three repository types to consider:
* 'containers' - 'standalone' top level repos that need externals, but can not be used as an external
* 'simple' - repos that only function as externals and never as a top level repo
* 'mixed' - repos that can be a top level or an external depending on context.

Consider:
* cesm - container
    - CESM.cfg - standalone container config
* clm - mixed
    - CESM.cfg - standalone container config
    - CLM.cfg - simple config
* cism / pop - simple
    - CISM.cfg - simple config

So we are always using 'CESM' as the standalone externals description, and 'component name' as the 'simple' externals description.

If I understand what you are proposing, it would result in Externals.cfg sometimes being a standalone description file, sometimes a simple externals description file:
* cesm - container
    - Externals.cfg - standalone container config
* clm - mixed
    - Externals.cfg - standalone container config
    - CLM.cfg - simple config
* cism / pop - simple
    - Externals.cfg - simple config

That's confusing because different locations use the same file for different purposes. But it also looks odd from the CLM standalone perspective. What's the difference between Externals.cfg and CLM.cfg? Is it more intuitive that 'Externals.cfg' point to the CLM externals, fates and ptclm? Maybe we need a second generic name... SubExternals.cfg? But that is odd for cism....

* cesm - container
    - Externals.cfg - standalone container config
* clm - mixed
    - Externals.cfg - standalone container config
    - SubExternals.cfg - simple config
* cism / pop - simple
    - SubExternals.cfg - simple config

We could use StandaloneExternals.cfg for the container files (and default name), and Externals.cfg for the simple externals, but that starts to look odd from the CESM perspective.

* cesm - container
    - StandaloneExternals.cfg - standalone container config
* clm - mixed
    - StandaloneExternals.cfg - standalone container config
    - Externals.cfg - simple config
* cism / pop - simple
    - Externals.cfg - simple config

Another comment somewhere wondered if we could assign a default file for sub-externals instead of listing them explicitly in the externals description file.

If we decided we wanted to use a default, one way to it would be to use '${COMPONENT_NAME}.cfg' as the default file name. How would it work with the proposed rename where some components use Externals for standalone, and some use Externals for simple? You'd have to migrate to something like 'Externals.cfg/SubExternals.cfg' or 'StandaloneExternals.cfg/Externals.cfg' for that to work....

Note: Using a default externals description file is just an example. I do not think it is a good idea because:

1. Creating new externals descriptions files from scratch is rare, so you are not saving much real work.
2. Reading externals description files is common and using a default makes it difficult to understand which components also have externals and why certain externals are being pulled in. E.g. Why is clm/fates being pulled in, it's not in cesm/CESM.cfg?
3. It's no longer clear how do you over-ride the default sub-externals file? For example fates development may want to always point to FatesDevCLM.cfg from CESM.cfg, leaving CLM.cfg untouched for easy testing.

**Comment by billsacks** [comment]
Let me start at the end: I'm fine with not having a default for the sub-externals file name and instead requiring it to be explicit. (I raised this as a question elsewhere, but I'm fine with the answer that it needs to be explicit: your arguments make sense.)

Now let me clarify (sorry: I should have been more explicit about this initially): I have imagined that even the "simple" repos will actually need two externals files, and this is how @mnlevy1981 and I have set up POP (note that CISM is now "mixed"). Currently this looks something like:

1. `CESM.cfg`

```
[pop]
local_path = .
protocol = externals_only
externals = POP.cfg
required = True
```

2. `POP.cfg`: lists the marbl and cvmix externals

While it's kind of funny to have this bare-bones `CESM.cfg` in POP, I think this addresses some of the concerns you had about consistency. So what I'm saying is that `CESM.cfg` doesn't make a lot of sense in this context.

My vote would be for:

- `Externals.cfg` - takes the place of the current `CESM.cfg` as the default

- `[COMPONENT_NAME]_Externals.cfg` (e.g., `CTSM_Externals.cfg`, `POP_Externals.cfg`) - I think this naming makes sense because these are the externals that are inherently part of the given component.

@bandre-ucar @mnlevy1981 @ekluzek what do you think?

**Comment by gold2718** [comment]
I do not like the idea of having cfg files not be free to have a descriptive name.
Rather than force the name to be generic, how about a default file which just contains a text string with the default cfg file to load? Some possible names are:

config_default.txt
config_default
defaults
defaults.txt

with the last two leaving open the possibility that there may be other default settings for that repo.

**Comment by billsacks** [comment]
@bandre-ucar just going through open issues... based on the discussion, I think this should be tagged with the Version 1.0 milestone; do you agree?

**Comment by bandre-ucar** [comment]
I feel that the current implementation is the least confusing option presented and don't see any particular benefit to making this change. But I'm fine with any direction that the group wants. 

We can probably make this for v1 if desired. But given that I don't see a consensus on what alternative to use and that we are getting very close to release... I'd prefer not to make this change now and defer until a future release.

**Comment by billsacks** [comment]
@bandre-ucar I'm fine with deferring... I was suggesting this for v1 largely based on your earlier comment:

> I don't have a preference, beyond if we make this change, it should happen before v1.0

Switching later will likely require cross-component coordination so that we end up with a CESM tag that is consistent in this respect. A bit painful, but doable.

As for what we want here, I'll second my own vote at the bottom of https://github.com/NCAR/manage_externals/issues/52#issuecomment-355288715

I personally don't love @gold2718 's suggestion (https://github.com/NCAR/manage_externals/issues/52#issuecomment-355369750): It seems like this adds some complexity and an extra file in every repo, and I don't see enough value coming from it. We need a hard-coded default for some file, and to me it doesn't seem to matter much whether the hard-coded default is (a) `Externals.cfg` (taking the place of the current `CESM.cfg`) or (b) the name of the file that points to `Externals.cfg`. But maybe @gold2718 sees a reason that (b) is better that I'm not seeing.

**Comment by gold2718** [comment]
@billsacks, sorry about the lack of a love fest :)
Let me make one more pitch for the idea. First, the 'extra file' can be in manage_externals and be 'configured' for the host repo once after manage_externals is imported. It would look like this:

Add a new file: `<root>/manage_externals/manic/default.py` which comes from the master repo as:
```
#!/usr/bin/env python

"""
Module for local configuration of the default externals configuration file.
"""

default_externals = "CESM.cfg"
```
or maybe with `default_externals = "Externals.cfg"`

Then, in checkout.py, simply add the line:
`from manic.default import default_externals`
and change `default='CESM.cfg'` to `default=default_externals`.

Still no love?

**Comment by gold2718** [comment]
I suppose an even easier idea which needs no code changes is to simply do the edit of checkout.py to make the default be whatever you want it to be for your repo. The repo-specific default will show up in `--help` so I'm fine with that.

**Comment by mnlevy1981** [comment]
So I like the idea of changing the default to `Externals.cfg` and, for components in what @bandre-ucar calls the "mixed type" also use  `[COMPONENT_NAME]_Externals.cfg`; thus POP would just have `Externals.cfg` but CTSM would have `Externals.cfg` and `CTSM_Externals.cfg`

> I feel that the current implementation is the least confusing option presented and don't see any particular benefit to making this change. 

I think the above suggestion is less confusing than every component having `CESM.cfg` or changing the default in the component-local version of `manage_externals`.

I also like @gold2718's suggestion of letting each component have a text file pointing to the default cfg file... e.g. `.manage_externals.cfg` that would specify `POP.cfg` for POP, `CTSM.cfg` for CTSM, `CESM.cfg` for CESM, and so on (again with the default of `Externals.cfg` if `.manage_externals.cfg` is not found). I guess my only question on this would be where in the directory tree it would live... I could see arguments for either `$POPROOT` or `$POPROOT/externals/manage_externals` (i.e. top level of the component or parallel to the `checkout_externals` script)

**Comment by gold2718** [comment]
@mnlevy1981, do you see a new text file as preferable to just editing the default in checkout.py?

**Comment by billsacks** [comment]
@bandre-ucar @mvertens @katetc and I just talked about this. Our preference is to avoid the extra level of complexity of having another file specifying the default. Our preference is also to avoid using CESM anywhere in the name.

@katetc 's suggestion, which we liked, is to have the default be `ComponentExternals.cfg`. However, now that I write that, I realize that I'd actually prefer `ExternalComponents.cfg`. To me, that has a nicer ring and is less ambiguous, but I can see how others may feel the opposite. `@katec @mvertens @bandre-ucar and others, do you have a preference?

Then the convention would be that the "internal externals" (how's that for an oxymoron) would be named with something like `CTSM.cfg` or maybe `Externals_CTSM.cfg`. But that's just a convention that would not be enforced anywhere in the code, and each component could make its own decision there. For POP, it would be fine to *just* have a single `ComponentExternals.cfg` / `ExternalComponents.cfg` file.

@gold2718 and @mnlevy1981, are you okay with this?

**Comment by mnlevy1981** [comment]
@gold2718 -- I do, mainly because editing the default in `checkout.py` raises the possibility of conficts if the default (or a line of code near the default) changes in a future version in this repository. I'd prefer for the POP checkout or this tool to be identical to a specific sha here.

@billsacks -- `ExternalComponents.cfg` works for me (as does `ComponentExternals.cfg`, though I agree with you that the former is less ambiguous)

**Comment by bandre-ucar** [comment]
Closed by 5fc8709

The default externals description files is: `Externals.cfg` as of v0.8.0.
The recommend naming convention for sub-externals in a repo that is mixed use as an external or  standalone is `Externals_{name}.cfg`.
